server/src/main/java/org/elasticsearch/search/retriever/rankdoc/RankDocsAndScoreSortField.java

```diff
@@ -43,36 +63,36 @@ public FieldComparator<?> newComparator(String fieldname, int numHits, Pruning p
         });
     }
 
-    private static class RankDocsComparator extends NumericComparator<Integer> {
-        private final int[] values;
-        private final Map<Integer, Integer> rankDocMap;
-        private int topValue;
-        private int bottom;
+    private static class RankDocsComparator extends NumericComparator<Long> {
+        private final long[] values;
+        private final Map<Integer, Long> rankDocMap;
+        private long topValue;
+        private long bottom;
 
         private RankDocsComparator(int numHits, RankDoc[] rankDocs) {
-            super(NAME, Integer.MAX_VALUE, false, Pruning.NONE, Integer.BYTES);
-            this.values = new int[numHits];
-            this.rankDocMap = Arrays.stream(rankDocs).collect(Collectors.toMap(k -> k.doc, v -> v.rank));
+            super(NAME, Long.MAX_VALUE, false, Pruning.NONE, Integer.BYTES);
+            this.values = new long[numHits];
+            this.rankDocMap = Arrays.stream(rankDocs).collect(Collectors.toMap(k -> k.doc, v -> encodeRankAndScore(v.rank, v.score)));
         }
 
         @Override
         public int compare(int slot1, int slot2) {
-            return Integer.compare(values[slot1], values[slot2]);
+            return Integer.compare(decodeRank(values[slot1]), decodeRank(values[slot2]));
```

Comment:
You should be able to compare without decoding rank. Unless you don't use the score as a tie breaker?

Reply:
Yeap, you're right. Updated in a63dd95a0f9cc4a7c0eb1e5c495ed77d719f011f